### PR TITLE
Extend SubModelPartSkinDetectionProcess

### PR DESCRIPTION
**Pull request by miguelmaso**
**Description**
Now it is possible to generate the skin for the part not belonging to a submodelpart.

**Changelog**
- Add new face selector to the process

### CODE REVIEW

File: kratos/processes/sub_model_part_skin_detection_process.cpp [hunk 1]
```diff
@@ -51,6 +71,16 @@ SubModelPartSkinDetectionProcess<TDim>::SubModelPartSkinDetectionProcess(
             settings["selection_settings"]["sub_model_part_name"].GetString()
         );
     }
+    else if (settings["selection_criteria"].GetString() == "node_not_on_sub_model_part")
```
Comment (loumalouomega) [inline kratos/processes/sub_model_part_skin_detection_process.cpp]:
```suggestion
   } else if (settings["selection_criteria"].GetString() == "node_not_on_sub_model_part") {
```

File: kratos/processes/sub_model_part_skin_detection_process.cpp [hunk 2]
```diff
@@ -51,6 +71,16 @@ SubModelPartSkinDetectionProcess<TDim>::SubModelPartSkinDetectionProcess(
             settings["selection_settings"]["sub_model_part_name"].GetString()
         );
     }
```
Comment (loumalouomega) [inline kratos/processes/sub_model_part_skin_detection_process.cpp]:
```suggestion
```

File: kratos/processes/sub_model_part_skin_detection_process.cpp [hunk 3]
```diff
@@ -51,6 +71,16 @@ SubModelPartSkinDetectionProcess<TDim>::SubModelPartSkinDetectionProcess(
...
+    else if (settings["selection_criteria"].GetString() == "node_not_on_sub_model_part")
+    {
+        KRATOS_ERROR_IF_NOT(settings["selection_settings"].Has("sub_model_part_names"))
+        << "When using \"selection_criteria\" == \"node_not_on_sub_model_part\","
+        << " SubModelPartSkinDetectionProcess requires the name of the target SubModelParts,"
+        << " given as the \"sub_model_part_names\" string array argument." << std::endl;
+        mpFaceSelector = Kratos::make_shared<SelectIfOneNodeNotOnSubModelPart>(
+            settings["selection_settings"]["sub_model_part_names"].GetStringArray()
+        );
+    }
```
Comment (loumalouomega) [inline kratos/processes/sub_model_part_skin_detection_process.cpp]:
```suggestion
    } else {
```

File: kratos/processes/sub_model_part_skin_detection_process.cpp [hunk 4]
```diff
@@ -51,6 +71,16 @@ SubModelPartSkinDetectionProcess<TDim>::SubModelPartSkinDetectionProcess(
...
     else
```
Comment (loumalouomega) [inline kratos/processes/sub_model_part_skin_detection_process.cpp]:
```suggestion
```

File: kratos/processes/sub_model_part_skin_detection_process.cpp [hunk 5]
```diff
@@ -51,6 +71,16 @@ SubModelPartSkinDetectionProcess<TDim>::SubModelPartSkinDetectionProcess(
...
     {
```
Comment (loumalouomega) [inline kratos/processes/sub_model_part_skin_detection_process.cpp]:
```suggestion
```

File: kratos/processes/sub_model_part_skin_detection_process.cpp [hunk 6]
```diff
@@ -50,9 +70,16 @@ SubModelPartSkinDetectionProcess<TDim>::SubModelPartSkinDetectionProcess(
         mpFaceSelector = Kratos::make_shared<SelectIfAllNodesOnSubModelPart>(
             settings["selection_settings"]["sub_model_part_name"].GetString()
         );
-    }
-    else
+   } else if (settings["selection_criteria"].GetString() == "node_not_on_sub_model_part") {
     {
```
Comment (loumalouomega) [inline kratos/processes/sub_model_part_skin_detection_process.cpp]:
```suggestion
```

File: kratos/processes/sub_model_part_skin_detection_process.h [hunk 7]
```diff
@@ -60,6 +60,19 @@ bool IsSelected(const Geometry<Node<3>>::PointsArrayType& rNodes) const override
 
 };
 
+// Select faces where almost one node does not belong to given SubModelPart.
+class SelectIfOneNodeNotOnSubModelPart: public FaceSelector
+{
+std::vector<std::string> mNames;
```
Comment (loumalouomega) [inline kratos/processes/sub_model_part_skin_detection_process.h]: Add some more detailed description Doxygen and tabulate the interior of the class

**Review by loumalouomega — CHANGES_REQUESTED** [review]
Minor comments

File: kratos/processes/sub_model_part_skin_detection_process.h (hunk 7 above)
Comment (miguelmaso) [reply kratos/processes/sub_model_part_skin_detection_process.h]: I've added some tabs and doxygen to the existing code.

File: kratos/processes/sub_model_part_skin_detection_process.h [hunk 8]
```diff
@@ -28,49 +28,66 @@ namespace Kratos
...
 ///@{
 
-/// Create a SubModelPart covering a part of the outside skin of the computation domain where a condition is met.
-/** For example, create the outer skin for the part of the domain belonging to a given SubModelPart.
-*/
+/**
+ * @class SubModelPartSkinDetectionProcess
+ * @brief Create a SubModelPart covering a part of the outside skin of the computation domain where a condition is met.
+ * @details For example, create the outer skin for the part of the domain belonging to a given SubModelPart.
+ * @author Jordi Cotela
```
Comment (loumalouomega) [inline kratos/processes/sub_model_part_skin_detection_process.h]: I think this is @rubenzorrilla

**Review by loumalouomega — DISMISSED** [review]

File: kratos/processes/sub_model_part_skin_detection_process.h (hunk 7 above)
Comment (miguelmaso) [reply kratos/processes/sub_model_part_skin_detection_process.h]: > tabulate the interior of the class

Do I continue tabulating the entire existing class?

File: kratos/processes/sub_model_part_skin_detection_process.h (hunk 8 above)
Comment (miguelmaso) [reply kratos/processes/sub_model_part_skin_detection_process.h]: let's wait @rubenzorrilla , I copied the header of the file

**Comment by miguelmaso** [comment]
@loumalouomega can you re-approve, please? The authorship is in the header of the file

**Review by loumalouomega — APPROVED** [review]

**Comment by miguelmaso** [comment]
thanks

**Pull request merged**